Library/src/Multiplayer/SignalRMsgPackEntitySerialiser.cpp

```diff
@@ -27,6 +27,178 @@ namespace csp::multiplayer
 {
 using namespace msgpack_typeids;
 
+namespace
+{
+std::pair<ItemComponentData, signalr::value> ReplicatedValueToSignalRValue(const ReplicatedValue& Value)
+{
+	ItemComponentData ValueType;
+	signalr::value NewValue;
+
+	switch (Value.GetReplicatedValueType())
+	{
+		case ReplicatedValueType::Boolean:
+			ValueType = ItemComponentData::BOOL;
+			NewValue  = signalr::value(Value.GetBool());
+			break;
+		case ReplicatedValueType::Integer:
+			ValueType = ItemComponentData::INT64;
+			NewValue  = signalr::value(Value.GetInt());
+			break;
+		case ReplicatedValueType::Float:
+			ValueType = ItemComponentData::FLOAT;
+			NewValue  = signalr::value(Value.GetFloat());
+			break;
+		case ReplicatedValueType::String:
+			ValueType = ItemComponentData::STRING;
+			NewValue  = signalr::value(Value.GetString().c_str(), Value.GetString().Length());
+			break;
+		case ReplicatedValueType::Vector2:
+		{
+			ValueType	= ItemComponentData::FLOAT_ARRAY;
+			auto VValue = Value.GetVector2();
+			NewValue	= signalr::value(std::vector {signalr::value(VValue.X), signalr::value(VValue.Y)});
+			break;
+		}
+		case ReplicatedValueType::Vector3:
+		{
+			ValueType	= ItemComponentData::FLOAT_ARRAY;
+			auto VValue = Value.GetVector3();
+			NewValue	= signalr::value(std::vector {signalr::value(VValue.X), signalr::value(VValue.Y), signalr::value(VValue.Z)});
+			break;
+		}
+		case ReplicatedValueType::Vector4:
+		{
+			ValueType	= ItemComponentData::FLOAT_ARRAY;
+			auto VValue = Value.GetVector4();
+			NewValue	= signalr::value(
+				   std::vector {signalr::value(VValue.X), signalr::value(VValue.Y), signalr::value(VValue.Z), signalr::value(VValue.W)});
+			break;
+		}
+		case ReplicatedValueType::Map:
+		{
+			ValueType = ItemComponentData::GENERIC_MAP;
+
+			// Creates a custom signalr data structure to hold an array of key-value pairs
+			const auto& Map = Value.GetMap();
+			auto Keys		= Map.Keys();
+
+			// Root of map which contains:
+			// 0 = Type information of the key
+			// 1 = Type information of value
+			// 2 = vector of key-value pairs
+			std::vector<signalr::value> MapRoot;
+			MapRoot.reserve(3);
+
+			// Array of key-value pairs
+			std::vector<signalr::value> MapArray;
+			MapArray.reserve(Keys->Size());
+
+			ItemComponentData KeyType	= ItemComponentData::UINT64;
+			ItemComponentData ValueType = ItemComponentData::UINT64;
+			bool First					= true;
+
+			for (size_t i = 0; i < Keys->Size(); ++i)
+			{
+				const ReplicatedValue& Key = (*Keys)[i];
+
+				auto KeyValue = ReplicatedValueToSignalRValue(Key);
+				auto Value	  = ReplicatedValueToSignalRValue(Map[Key]);
+
+				if (First)
+				{
+					First	  = false;
+					KeyType	  = KeyValue.first;
+					ValueType = Value.first;
+				}
+				else
+				{
+					// Ensure key and values are all the same type
+					assert(KeyType == KeyValue.first && "Key types dont match!");
```

Comment:
No strong preference, more curious - why assert instead of logging an error?

Reply:
I was following the convention of this file, which always asserts (except 1 instance)

Reply:
Do you think it's the right one to follow here? (Not a leading question)

Reply:
Thinking about it more I think it makes sense to error, if clients could setup the map incorrectly,  I think they should be notified through an error.

Reply:
Updated!